### PR TITLE
Homework1 - Stroke prediction classification task

### PR DESCRIPTION
**Pull request by Tais-S**
Данные для классификации представляют собой различные параметры людей (как численные, так и категориальные), на основе которых делается вывод о том, предрасположен ли некоторый человек инсульту. Так как в данных имеется сильная несбалансированность по классу, то перед тренировкой модели проводится oversampling, то есть дополнение данных наименьшим классом. В качестве наилучшей модели из протестированных (модели из sklearn) предлагается использовать SVC с параметрами C=100, gamma=0.001, kernel='sigmoid'. Для оценки модели особое внимание было уделено метрике false_negative_rate, чтобы минимизировать число случаев, когда человеку с большой вероятностью инсульта проставляется негативный класс.

Самооценка:
0) Описание решений в пул-реквесте (1/1)
1) Самооценка в пул-реквесте (1/1)
2) Ноутбук с EDA (1/1)
3) Функция train, вызов утилитой, прописано в readme (2.5/3)
4) Функция predict, вызов утилитой, прописано в readme (2.5/3)
(Сняла по полбалла за то, что вызов утилитой работает только если немного сменить импорт функций, а именно, избавиться от прокладывания пути, начиная с 'ml_project.'. В противном случае не работают тесты. Долго пыталась решить проблему, в частности, выставлением source directory и working directory, как советуют в интернетах, но все же не смогла это исправить)
5) Модулность (2/2)
6) Логгеры (2/2)
7) Тесты на модули, на трейн и на предикт (3/3)
8) Синтетические данные (2/2)
9) 2 конфигурации (3/3)
10) Датаклассы (2/2)
11) Трансформер с тестом (3/3)
12) Фиксация зависимостей (1/1)
13) CI (3/3)

Итого: 29/30

### CODE REVIEW

File: tests/generate_data.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        generated_col = generate_column(data, col_name, n+1, float_distributions)
+        synthetic_data[col_name] = generated_col
+    synthetic_data_train, synthetic_data_test = train_test_split(synthetic_data,
+                                                                 test_size=0.2,
+                                                                 random_state=2022)
+    synthetic_data_test = synthetic_data_test.drop("stroke", 1)
+    with open("tests/synthetic_data_train.csv", "wb") as f:
+        synthetic_data_train.to_csv(f)
+    with open("tests/synthetic_data_test.csv", "wb") as f:
+        synthetic_data_test.to_csv(f)
```
Comment (ningeen) [inline tests/generate_data.py:44]: Лучше было воспользоваться fixture https://docs.pytest.org/en/6.2.x/tmpdir.html

File: tests/generate_data.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    return synthetic_data_train, synthetic_data_test
+
+
+@click.command(name="generate_data")
+@click.argument("n", type=int)
+def generate_data_command(n):
+    click.echo("started...")
+    data = pd.read_csv("data/raw/train_df.csv")
```
Comment (ningeen) [inline tests/generate_data.py:52]: Не стоит хардкодить пути к файлам

File: ml_project/predict_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+def run_predict_pipeline(predict_pipeline_params):
+    with open(predict_pipeline_params.output_model_path, "rb") as f:
+        model = pickle.load(f)
+    with open(predict_pipeline_params.transformer_path, "rb") as f:
+        transformer = pickle.load(f)
+    test_df = read_data(predict_pipeline_params.test_data_path)
+    test_df = test_df.drop(predict_pipeline_params.feature_params.features_to_drop, 1)
+    logger.info(f"===== test columns {test_df.columns}")
+    inference_pipeline = create_inference_pipeline(model, transformer)
+    predictions = predict_model(model, test_df)
```
Comment (ningeen) [inline ml_project/predict_pipeline.py:31]: Тут не баг случаем? inference_pipeline создается, но не используется

File: ml_project/models/model_predict.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from sklearn.pipeline import Pipeline
+
+
+def predict_model(model, features):
+    predictions = model.predict(features)
+    return predictions
+
+
+def create_inference_pipeline(model, transformer):
+    return Pipeline([("features_part", transformer), ("model_part", model)])
```
Comment (ningeen) [inline ml_project/models/model_predict.py:10]: Все эти функции есть в `model_fit.py` Зачем дублировать?

**Review by ningeen — COMMENTED** [review]
Замечания:
1. Данные лучше не хранить в гите, а подтягивать с удаленного хранилища
1. Под кастомным трансформером имелось в виду создать свой класс, отнаследовавшись от BaseEstimator and TransformerMixin


Итого 27 баллов
1. В описании к пулл реквесту описаны основные "архитектурные" и тактические решения, которые сделаны в вашей работе. В общем, описание того, что именно вы сделали и для чего, чтобы вашим ревьюерам было легче понять ваш код (1/1)
1. В пулл-реквесте проведена самооценка, распишите по каждому пункту выполнен ли критерий или нет и на сколько баллов(частично или полностью) (1/1)
1. Выполнено EDA, закоммитьте ноутбук в папку с ноутбуками (1/1)
1. Написана функция/класс для тренировки модели, вызов оформлен как утилита командной строки, записана в readme инструкцию по запуску (3/3)
1. Написана функция/класс predict (вызов оформлен как утилита командной строки), которая примет на вход артефакт/ы от обучения, тестовую выборку (без меток) и запишет предикт по заданному пути, инструкция по вызову записана в readme (3/3)
1. Проект имеет модульную структуру (2/2)
1. Использованы логгеры (2/2)
1. Написаны тесты на отдельные модули и на прогон обучения и predict (3/3)
1. Для тестов генерируются синтетические данные, приближенные к реальным (2/2)
1. Обучение модели конфигурируется с помощью конфигов в json или yaml, закоммитьте как минимум 2 корректные конфигурации, с помощью которых можно обучить модель (разные модели, стратегии split, preprocessing) (3/3)
1. Используются датаклассы для сущностей из конфига, а не голые dict (2/2)
1. Напишите кастомный трансформер и протестируйте его (0/3)
1. В проекте зафиксированы все зависимости (1/1)
1. Настроен CI для прогона тестов, линтера на основе github actions (3/3).

**Comment by Tais-S** [comment]
Спасибо за проверку! Можно поставить аппрув на мой пул-реквест?

**Review by ningeen — APPROVED** [review]
